File: src/main/java/com/freya02/botcommands/api/prefixed/annotations/TextOption.java
package com.freya02.botcommands.api.prefixed.annotations;

import com.freya02.botcommands.api.annotations.Optional;
import org.jetbrains.annotations.Nullable;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Annotation used to specify a text command parameter is supplied from a Discord message, <i>i.e. it is <b>not</b> a custom parameter</i>
 * <br>This also can set name and example of {@linkplain JDATextCommand text commands} parameters
 * <p>
 * {@linkplain #name()} is optional if the parameter name is available (add -parameters to your java compiler)
 *
 * @see Optional Optional (can also see @Nullable)
 * @see Nullable
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.PARAMETER})
public @interface TextOption {
	String name() default "";

	String example() default "";
}

File: BotCommands-commands/text/src/main/kotlin/io/github/freya022/botcommands/api/commands/text/annotations/TextOption.kt
package io.github.freya022.botcommands.api.commands.text.annotations

import io.github.freya022.botcommands.api.commands.annotations.Optional
import io.github.freya022.botcommands.api.commands.annotations.VarArgs
import io.github.freya022.botcommands.api.parameters.resolvers.TextParameterResolver
import org.jspecify.annotations.Nullable

/**
 * Sets a parameter as a text command option from the Discord message.
 *
 * This also can set name and example of [text commands][JDATextCommandVariation] parameters.
 *
 * The supported data types can be seen in [TextParameterResolver], more types can be supported by implementing it.
 *
 * ### Varargs
 *
 * You can define a single vararg parameter per text command variation, using [@VarArgs][VarArgs].
 *
 * @see Optional @Optional
 * @see Nullable @Nullable
 * @see ID @ID
 */
@Target(AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
annotation class TextOption(
    /**
     * The name of this option displayed on the help content.
     *
     * This is optional if the parameter name is available,
     * see [the wiki](https://bc.freya02.dev/3.X/setup/parameter-names/) for more details.
     */
    val name: String = "",

    /**
     * The example input of this option displayed on the help content.
     */
    val example: String = ""
)
